src/printer.js

```diff
@@ -152,24 +130,138 @@ function isLastStatement(path) {
   return body[body.length - 1] === node;
 }
 
-function printLines(path, options, print) {
-  const text = options.originalText;
-  const printed = path.map(stmtPath => {
-    const stmt = stmtPath.getValue();
-    const parts = [];
-    parts.push(print(stmtPath));
-    if (lineShouldEndWithSemicolon(stmtPath)) {
-      parts.push(";");
-    }
-    if (!isLastStatement(stmtPath)) {
-      if (sharedUtil.isNextLineEmpty(text, stmt, options)) {
-        parts.push(hardline);
-      }
-      parts.push(hardline);
+function nodeHasStatement(node) {
+  return [
+    "block",
+    "program",
+    "namespace",
+    "class",
+    "interface",
+    "trait",
+    "traituse",
+    "declare"
+  ].includes(node.kind);
+}
+
+function isFirstNodeInParentProgramNode(path) {
+  const parentNode = path.getParentNode();
+  const nodeIndex = getNodeIndex(path);
+  const isParentProgramNode = parentNode && parentNode.kind === "program";
+  return isParentProgramNode && nodeIndex === 0;
+}
+
+function isFirstNodeInParentNode(path) {
+  const nodeIndex = getNodeIndex(path);
+  return nodeIndex === 0;
+}
+
+function isLastNodeInParentNode(path) {
+  const parentNodeBody = getParentNodeListProperty(path);
+  const nodeIndex = getNodeIndex(path);
+  return parentNodeBody && nodeIndex === parentNodeBody.length - 1;
+}
+
+function isPrevNodeInline(path) {
+  const nodeIndex = getNodeIndex(path);
+  const parentNodeBody = getParentNodeListProperty(path);
+  const prevNode = nodeIndex !== -1 && parentNodeBody[nodeIndex - 1];
+  return prevNode && prevNode.kind === "inline";
+}
+
+function isNextNodeInline(path) {
+  const parentNodeBody = getParentNodeListProperty(path);
+  const nodeIndex = getNodeIndex(path);
+  const nextNode = nodeIndex !== -1 && parentNodeBody[nodeIndex + 1];
+  return nextNode && nextNode.kind === "inline";
+}
+
+function shouldRemoveRemoveLines(path) {
+  return isPrevNodeInline(path) && isNextNodeInline(path);
+}
+
+function lineShouldEndWithHardline(path) {
+  const node = path.getValue();
+  const nodeIndex = getNodeIndex(path);
+  return (
+    nodeIndex !== -1 &&
+    !isLastStatement(path) &&
+    node.kind !== "case" &&
+    !isNextNodeInline(path)
+  );
+}
+
+function lineShouldHaveStartPHPTag(path) {
+  return isFirstNodeInParentProgramNode(path) || isPrevNodeInline(path);
+}
+
+function lineShouldHaveEndPHPTag(path) {
+  return isNextNodeInline(path);
+}
+
+function fileShouldEndWithHardline(path) {
```

Comment:
is there a reason this should be separate from `lineShouldEndWithHardline`?

Reply:
@mgrip `lineShouldEndWithHardline` for statement, `fileShouldEndWithHardline` for `program` node - i.e. end of file